File: Data Structures & Algorithms/binary-tree-right-side-view/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        out = []
        maxLevel = 0
        currLevel = 1
        def dfs(node, level):
            nonlocal maxLevel
            if not node:
                return
            if level > maxLevel:
                out.append(node.val)
                maxLevel = level
            dfs(node.right, level + 1)
            dfs(node.left, level + 1)
        dfs(root, currLevel)
        return out

File: Data Structures & Algorithms/balanced-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int maxDepth(TreeNode* root) {
        if (root == nullptr) {
            return 0;
        }
        int left = 1 + maxDepth(root->left);
        int right = 1 + maxDepth(root->right);
        if (left > right) {
            return left;
        } else {
            return right;
        }
    }
    bool isBalanced(TreeNode* root) {
        if (!root) {
            return true;
        }
        if (abs(maxDepth(root->left) - maxDepth(root->right)) > 1) {
            return false;
        }
        return (isBalanced(root->left) && isBalanced(root->right));
    }
};
